### ISSUE TITLE
评估报错

### ISSUE THREAD

**Issue by htired** [WujiangXu/SLMRec#7]
![Image](https://github.com/user-attachments/assets/dbdb291f-3a1d-4cc7-97e5-5e39ddc8c384)
好像缺少模型训练的步骤，能补充一下这部分吗？
此外，我已添加 trainer.train(resume_from_checkpoint=resume_from_checkpoint) 以支持从中断处恢复训练，
但在执行**评估**时仍然遇到了错误：
![Image](https://github.com/user-attachments/assets/c8746fca-db1e-406d-b988-f7743e93bcf2)

**Comment by WujiangXu** [comment]
Hi, thank you for the attention. Could you provide more information to reproduce the bug?

**Comment by htired** [comment]
> Hi, thank you for the attention. Could you provide more information to reproduce the bug?

---
First，当我没有加 trainer.train(resume_from_checkpoint=resume_from_checkpoint)
![Image](https://github.com/user-attachments/assets/83f83df1-8cff-4d68-9c7c-42e14c760898)
---
Second：pred_out = trainer.predict(test_dataset=datasetTest)
当我加了trainer.train(resume_from_checkpoint=resume_from_checkpoint)，会出现如下错误。
def compute_metrics(pred):
    """Compute all evaluation metrics for a prediction.
    
    Args:
        pred: Prediction object containing logits and labels
        
    Returns:
        Dictionary of computed metrics
    """
    logits = pred.predictions
    labels = pred.label_ids
    metrics = {}
    print("pred.predictions",pred.predictions) # [[], []]
    print("pred.label_ids",pred.label_ids) # [0 1 2 3 ...]
    print("pred.label_ids[0]",pred.label_ids[0]) # 0
    # Calculate metrics at different K values
    for k in [1, 5, 10]:
        r = getLabel(labels, logits)
        metrics[f'ndcg_{k}'] = NDCG_atK(labels, r, k)
        metrics[f'hit_{k}'] = RecallPrecision_atK(labels, r, k)[1]
    
    metrics['mrr'] = MRR_atK(labels, r, 10)
    return metrics
------
def getLabel(test, pred):
    """Convert test and predictions to binary relevance matrix.
    
    Args:
        test: Ground truth items
        pred: Predicted items
        
    Returns:
        Binary relevance matrix
    """
    r = []
    for i in range(len(test)):
        groundTrue = test[i]
        predictTopK = pred[i]
        pred_list = list(map(lambda x: x in groundTrue, predictTopK))
        pred_list = np.array(pred_list).astype("float")
        r.append(pred_list)
    return np.array(r).astype('float')
---
传进去的 test 就是 0。我修改为 pred.label_ids 也出现了问题。

**Comment by htired** [comment]
> > Hi, thank you for the attention. Could you provide more information to reproduce the bug?
> 
> ## First，当我没有加 trainer.train(resume_from_checkpoint=resume_from_checkpoint)
> ![Image](https://github.com/user-attachments/assets/83f83df1-8cff-4d68-9c7c-42e14c760898)
> Second：pred_out = trainer.predict(test_dataset=datasetTest) 当我加了trainer.train(resume_from_checkpoint=resume_from_checkpoint)，会出现如下错误。 def compute_metrics(pred): """Compute all evaluation metrics for a prediction.
> 
> ```
> Args:
>     pred: Prediction object containing logits and labels
>     
> Returns:
>     Dictionary of computed metrics
> """
> logits = pred.predictions
> labels = pred.label_ids
> metrics = {}
> print("pred.predictions",pred.predictions) # [[], []]
> print("pred.label_ids",pred.label_ids) # [0 1 2 3 ...]
> print("pred.label_ids[0]",pred.label_ids[0]) # 0
> # Calculate metrics at different K values
> for k in [1, 5, 10]:
>     r = getLabel(labels, logits)
>     metrics[f'ndcg_{k}'] = NDCG_atK(labels, r, k)
>     metrics[f'hit_{k}'] = RecallPrecision_atK(labels, r, k)[1]
> 
> metrics['mrr'] = MRR_atK(labels, r, 10)
> return metrics
> ```
> 
> def getLabel(test, pred): """Convert test and predictions to binary relevance matrix.
> 
> ```
> Args:
>     test: Ground truth items
>     pred: Predicted items
>     
> Returns:
>     Binary relevance matrix
> """
> r = []
> for i in range(len(test)):
>     groundTrue = test[i]
>     predictTopK = pred[i]
>     pred_list = list(map(lambda x: x in groundTrue, predictTopK))
>     pred_list = np.array(pred_list).astype("float")
>     r.append(pred_list)
> return np.array(r).astype('float')
> ```
> 
> 传进去的 test 就是 0。我修改为 pred.label_ids 也出现了问题。

**Comment by WujiangXu** [comment]
> > Hi, thank you for the attention. Could you provide more information to reproduce the bug?
> 
> ## First，当我没有加 trainer.train(resume_from_checkpoint=resume_from_checkpoint)
> ![Image](https://github.com/user-attachments/assets/83f83df1-8cff-4d68-9c7c-42e14c760898)
> Second：pred_out = trainer.predict(test_dataset=datasetTest) 当我加了trainer.train(resume_from_checkpoint=resume_from_checkpoint)，会出现如下错误。 def compute_metrics(pred): """Compute all evaluation metrics for a prediction.
> 
> ```
> Args:
>     pred: Prediction object containing logits and labels
>     
> Returns:
>     Dictionary of computed metrics
> """
> logits = pred.predictions
> labels = pred.label_ids
> metrics = {}
> print("pred.predictions",pred.predictions) # [[], []]
> print("pred.label_ids",pred.label_ids) # [0 1 2 3 ...]
> print("pred.label_ids[0]",pred.label_ids[0]) # 0
> # Calculate metrics at different K values
> for k in [1, 5, 10]:
>     r = getLabel(labels, logits)
>     metrics[f'ndcg_{k}'] = NDCG_atK(labels, r, k)
>     metrics[f'hit_{k}'] = RecallPrecision_atK(labels, r, k)[1]
> 
> metrics['mrr'] = MRR_atK(labels, r, 10)
> return metrics
> ```
> 
> def getLabel(test, pred): """Convert test and predictions to binary relevance matrix.
> 
> ```
> Args:
>     test: Ground truth items
>     pred: Predicted items
>     
> Returns:
>     Binary relevance matrix
> """
> r = []
> for i in range(len(test)):
>     groundTrue = test[i]
>     predictTopK = pred[i]
>     pred_list = list(map(lambda x: x in groundTrue, predictTopK))
>     pred_list = np.array(pred_list).astype("float")
>     r.append(pred_list)
> return np.array(r).astype('float')
> ```
> 
> 传进去的 test 就是 0。我修改为 pred.label_ids 也出现了问题。

It may be caused by the version of some packages. Check the requirements file.

**Comment by htired** [comment]
> > > Hi, thank you for the attention. Could you provide more information to reproduce the bug?
> > 
> > 
> > ## First，当我没有加 trainer.train(resume_from_checkpoint=resume_from_checkpoint)
> > ![Image](https://github.com/user-attachments/assets/83f83df1-8cff-4d68-9c7c-42e14c760898)
> > Second：pred_out = trainer.predict(test_dataset=datasetTest) 当我加了trainer.train(resume_from_checkpoint=resume_from_checkpoint)，会出现如下错误。 def compute_metrics(pred): """Compute all evaluation metrics for a prediction.
> > ```
> > Args:
> >     pred: Prediction object containing logits and labels
> >     
> > Returns:
> >     Dictionary of computed metrics
> > """
> > logits = pred.predictions
> > labels = pred.label_ids
> > metrics = {}
> > print("pred.predictions",pred.predictions) # [[], []]
> > print("pred.label_ids",pred.label_ids) # [0 1 2 3 ...]
> > print("pred.label_ids[0]",pred.label_ids[0]) # 0
> > # Calculate metrics at different K values
> > for k in [1, 5, 10]:
> >     r = getLabel(labels, logits)
> >     metrics[f'ndcg_{k}'] = NDCG_atK(labels, r, k)
> >     metrics[f'hit_{k}'] = RecallPrecision_atK(labels, r, k)[1]
> > 
> > metrics['mrr'] = MRR_atK(labels, r, 10)
> > return metrics
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > def getLabel(test, pred): """Convert test and predictions to binary relevance matrix.
> > ```
> > Args:
> >     test: Ground truth items
> >     pred: Predicted items
> >     
> > Returns:
> >     Binary relevance matrix
> > """
> > r = []
> > for i in range(len(test)):
> >     groundTrue = test[i]
> >     predictTopK = pred[i]
> >     pred_list = list(map(lambda x: x in groundTrue, predictTopK))
> >     pred_list = np.array(pred_list).astype("float")
> >     r.append(pred_list)
> > return np.array(r).astype('float')
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 传进去的 test 就是 0。我修改为 pred.label_ids 也出现了问题。
> 
> It may be caused by the version of some packages. Check the requirements file.

我使用了您给的环境，还是出现了相同的问题。

**Comment by WujiangXu** [comment]
> > > > Hi, thank you for the attention. Could you provide more information to reproduce the bug?
> > > 
> > > 
> > > ## First，当我没有加 trainer.train(resume_from_checkpoint=resume_from_checkpoint)
> > > ![Image](https://github.com/user-attachments/assets/83f83df1-8cff-4d68-9c7c-42e14c760898)
> > > Second：pred_out = trainer.predict(test_dataset=datasetTest) 当我加了trainer.train(resume_from_checkpoint=resume_from_checkpoint)，会出现如下错误。 def compute_metrics(pred): """Compute all evaluation metrics for a prediction.
> > > ```
> > > Args:
> > >     pred: Prediction object containing logits and labels
> > >     
> > > Returns:
> > >     Dictionary of computed metrics
> > > """
> > > logits = pred.predictions
> > > labels = pred.label_ids
> > > metrics = {}
> > > print("pred.predictions",pred.predictions) # [[], []]
> > > print("pred.label_ids",pred.label_ids) # [0 1 2 3 ...]
> > > print("pred.label_ids[0]",pred.label_ids[0]) # 0
> > > # Calculate metrics at different K values
> > > for k in [1, 5, 10]:
> > >     r = getLabel(labels, logits)
> > >     metrics[f'ndcg_{k}'] = NDCG_atK(labels, r, k)
> > >     metrics[f'hit_{k}'] = RecallPrecision_atK(labels, r, k)[1]
> > > 
> > > metrics['mrr'] = MRR_atK(labels, r, 10)
> > > return metrics
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > def getLabel(test, pred): """Convert test and predictions to binary relevance matrix.
> > > ```
> > > Args:
> > >     test: Ground truth items
> > >     pred: Predicted items
> > >     
> > > Returns:
> > >     Binary relevance matrix
> > > """
> > > r = []
> > > for i in range(len(test)):
> > >     groundTrue = test[i]
> > >     predictTopK = pred[i]
> > >     pred_list = list(map(lambda x: x in groundTrue, predictTopK))
> > >     pred_list = np.array(pred_list).astype("float")
> > >     r.append(pred_list)
> > > return np.array(r).astype('float')
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > 传进去的 test 就是 0。我修改为 pred.label_ids 也出现了问题。
> > 
> > 
> > It may be caused by the version of some packages. Check the requirements file.
> 
> 我使用了您给的环境，还是出现了相同的问题。

Have you ever solved the problem? If not, I will re-run the experiment next week. Besides, could you share the version of pip package?

**Comment by htired** [comment]
> > > > > Hi, thank you for the attention. Could you provide more information to reproduce the bug?
> > > > 
> > > > 
> > > > ## First，当我没有加 trainer.train(resume_from_checkpoint=resume_from_checkpoint)
> > > > ![Image](https://github.com/user-attachments/assets/83f83df1-8cff-4d68-9c7c-42e14c760898)
> > > > Second：pred_out = trainer.predict(test_dataset=datasetTest) 当我加了trainer.train(resume_from_checkpoint=resume_from_checkpoint)，会出现如下错误。 def compute_metrics(pred): """Compute all evaluation metrics for a prediction.
> > > > ```
> > > > Args:
> > > >     pred: Prediction object containing logits and labels
> > > >     
> > > > Returns:
> > > >     Dictionary of computed metrics
> > > > """
> > > > logits = pred.predictions
> > > > labels = pred.label_ids
> > > > metrics = {}
> > > > print("pred.predictions",pred.predictions) # [[], []]
> > > > print("pred.label_ids",pred.label_ids) # [0 1 2 3 ...]
> > > > print("pred.label_ids[0]",pred.label_ids[0]) # 0
> > > > # Calculate metrics at different K values
> > > > for k in [1, 5, 10]:
> > > >     r = getLabel(labels, logits)
> > > >     metrics[f'ndcg_{k}'] = NDCG_atK(labels, r, k)
> > > >     metrics[f'hit_{k}'] = RecallPrecision_atK(labels, r, k)[1]
> > > > 
> > > > metrics['mrr'] = MRR_atK(labels, r, 10)
> > > > return metrics
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > def getLabel(test, pred): """Convert test and predictions to binary relevance matrix.
> > > > ```
> > > > Args:
> > > >     test: Ground truth items
> > > >     pred: Predicted items
> > > >     
> > > > Returns:
> > > >     Binary relevance matrix
> > > > """
> > > > r = []
> > > > for i in range(len(test)):
> > > >     groundTrue = test[i]
> > > >     predictTopK = pred[i]
> > > >     pred_list = list(map(lambda x: x in groundTrue, predictTopK))
> > > >     pred_list = np.array(pred_list).astype("float")
> > > >     r.append(pred_list)
> > > > return np.array(r).astype('float')
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > 传进去的 test 就是 0。我修改为 pred.label_ids 也出现了问题。
> > > 
> > > 
> > > It may be caused by the version of some packages. Check the requirements file.
> > 
> > 
> > 我使用了您给的环境，还是出现了相同的问题。
> 
> Have you ever solved the problem? If not, I will re-run the experiment next week. Besides, could you share the version of pip package?

没有，我使用了您给的 pip package 版本

![Image](https://github.com/user-attachments/assets/d73ac355-f8e3-48c7-a126-621a5306e309)

**Comment by WujiangXu** [comment]
> I also encountered the same problem. Could you please tell me if you have any solutions?

Hi, I will check again in a few days.

**Comment by wcx12** [comment]
I'm not sure if my solution is correct.
First, you need to add the training code at this location in finetune.py. Otherwise, the model will skip the training phase and directly proceed to the testing phase.

![Image](https://github.com/user-attachments/assets/c62365cf-06aa-4f51-8787-e704de0f4495)

Then, during the testing phase, you need to modify the function compute_metrics. Because I found that logits = pred.predictions and labels = pred.label_ids seem to be inconsistent with the format of the function you called later. Here, logits actually represents num_user * (1 + the number of negative samples), meaning that the first element of each row represents the predicted value of the positive sample.
Therefore, I implemented a function myself to calculate the evaluation indicators.

![Image](https://github.com/user-attachments/assets/e6acfa94-a8de-4479-b8b7-c0ecdb7bd16e)

![Image](https://github.com/user-attachments/assets/aacb1671-dc22-4863-aed1-c1eb5648b919)

**Comment by WujiangXu** [comment]
> I'm not sure if my solution is correct. First, you need to add the training code at this location in finetune.py. Otherwise, the model will skip the training phase and directly proceed to the testing phase.
> 
> ![Image](https://github.com/user-attachments/assets/c62365cf-06aa-4f51-8787-e704de0f4495)
> 
> Then, during the testing phase, you need to modify the function compute_metrics. Because I found that logits = pred.predictions and labels = pred.label_ids seem to be inconsistent with the format of the function you called later. Here, logits actually represents num_user * (1 + the number of negative samples), meaning that the first element of each row represents the predicted value of the positive sample. Therefore, I implemented a function myself to calculate the evaluation indicators.
> 
> ![Image](https://github.com/user-attachments/assets/e6acfa94-a8de-4479-b8b7-c0ecdb7bd16e)
> 
> ![Image](https://github.com/user-attachments/assets/aacb1671-dc22-4863-aed1-c1eb5648b919)

Your solution is correct. This line of code was overwritten during a previous update, and we have now resolved this issue. @htired 

Regarding the evaluation metrics, we follow the traditional evaluation metric computation approach, such as the one used in "https://github.com/kuandeng/LightGCN", which places the positive prediction as the first element. @wcx12

**Comment by WujiangXu** [comment]
@htired @wcx12 

Thank you for reporting this issue. I have reviewed the code and discovered that some critical parts of eval_utils.py were modified during the recent code cleanup. The branch 2fb46cc37ca34451612e36e8ac16bab138f3cf56 works successfully.

To resolve this bug, I have updated the code. Please check the latest version.

In the next update, I will clean the code carefully.
